src/Swift.Bindings/src/Model/TypeSpec.cs

```diff
@@ -0,0 +1,380 @@
+		ty!.IsInOut = false;
+		return ty!;
+	}
+
+	static TypeSpec ProjectAsNonReference(TypeSpec a)
+	{
+		if (a.IsInOut)
+		{
+			return a.NonReferenceCloneOf();
+		}
+		if (a is NamedTypeSpec namedType && namedType.GenericParameters.Count == 1)
+		{
+			if (namedType.Name == "Swift.UnsafePointer" || namedType.Name == "Swift.UnsafeMutablePointer")
+				return namedType.GenericParameters[0];
+		}
+		return a;
+	}
+
+	public override int GetHashCode()
+	{
+		return ToString().GetHashCode();
+	}
+
+	protected static bool ListEqual(List<TypeSpec> one, List<TypeSpec> two, bool partialNameMatch)
+	{
+		if (one.Count != two.Count)
+			return false;
+		for (int i = 0; i < one.Count; i++)
+		{
+			if (partialNameMatch)
+			{
+				if (!one[i].EqualsPartialMatch(two[i]))
+					return false;
+			}
+			else
+			{
+				if (!one[i].Equals(two[i]))
+					return false;
+			}
+		}
+		return true;
+	}
+
+	public static bool IsNullOrEmptyTuple(TypeSpec? spec)
+	{
+		return spec is null || spec.IsEmptyTuple;
+	}
+
+	public static bool BothNullOrEqual(TypeSpec? one, TypeSpec? two)
+	{
+		if (one is null && two is null)
+			return true;
+		if (one is null || two is null)
+			return false;
+		return one.Equals(two);
+	}
+
+	public bool ContainsBoundGenericClosure()
+	{
+		return ContainsBoundGenericClosure(0);
+	}
+
+	bool ContainsBoundGenericClosure(int depth)
+	{
+		if (this is NamedTypeSpec namedTypeSpec)
+		{
+			foreach (var subSpec in namedTypeSpec.GenericParameters)
+			{
+				if (subSpec.ContainsBoundGenericClosure(depth + 1))
+					return true;
+			}
+		}
+		else if (this is TupleTypeSpec tupleSpec)
+		{
+			foreach (var subSpec in tupleSpec.Elements)
+			{
+				if (subSpec.ContainsBoundGenericClosure(depth + 1))
+					return true;
+			}
+		}
+		else if (this is ClosureTypeSpec closureSpec)
+		{
+			return depth > 0;
+		}
+		return false;
+	}
+
+	public override string ToString()
+	{
+		return ToString(true);
+	}
+
+	public string ToString(bool useFullNames)
+	{
+		StringBuilder builder = new StringBuilder();
+
+		foreach (var attr in Attributes)
+		{
+			builder.Append(attr.ToString());
+			builder.Append(' ');
+		}
+		if (IsInOut)
+			builder.Append("inout ");
+
+		if (IsAny)
+			builder.Append("any ");
+
+		if (TypeLabel is not null)
+		{
+			builder.Append(TypeLabel).Append(": ");
+		}
+		builder.Append(LLToString(useFullNames));
+
+		if (ContainsGenericParameters)
+		{
+			builder.Append('<');
+			for (int i = 0; i < GenericParameters.Count; i++)
+			{
+				if (i > 0)
+					builder.Append(", ");
+				builder.Append(GenericParameters[i].ToString(useFullNames));
+			}
+			builder.Append('>');
+		}
+		builder.Append(LLFinalStringParts());
+
+		return builder.ToString();
+	}
+
+	static string[] intNames = {
+			"Swift.Int", "Swift.UInt", "Swift.Int8", "Swift.UInt8",
+			"Swift.Int16", "Swift.UInt16", "Swift.Int32", "Swift.UInt32",
+			"Swift.Int64", "Swift.UInt64", "Swift.Char"
+		};
+
+	public static bool IsIntegral(TypeSpec ts)
```

Comment:
Should these getters be in TypeDecl?

Reply:
It should probably be in the type database as `IsIntegral (TypeDecl td)` and `IsIntegral(TypeSpec ts)`. I don't see it used immediately, so I'll remove it entirely.